### PR TITLE
Unify the example programs' termination

### PR DESCRIPTION
**Pull request by k-stachowiak**
This is done to account for platforms, for which we want custom behavior
upon the program termination, hence we call `mbedtls_exit()` instead of
returning from `main()`.

This PR is intended to lighten the overblown #1572. If it get's in, the original PR's complexity may be reduced. I intend to create several PRs like this, based on the same reasoning.

Backports: #2705 #2710

### CODE REVIEW

**Comment by k-stachowiak** [comment]
@RonEld Thanks! I think it may be a good idea. These files weren't modified in the original PR as they aren't supposed to use platform resources. On the other hand, this PR is more general and I don't see any reason not to broaden its scope.

**Comment by k-stachowiak** [comment]
@RonEld I added the changes to other example programs for consistency. Can you have another look? I was trying to get a clean CI pass before the re-review, but it seems like it's having some internal issues.

File: programs/pkey/rsa_verify.c [hunk 1]
```diff
@@ -45,7 +45,7 @@ int main( void )
     mbedtls_printf("MBEDTLS_BIGNUM_C and/or MBEDTLS_RSA_C and/or "
             "MBEDTLS_MD_C and/or "
             "MBEDTLS_SHA256_C and/or MBEDTLS_FS_IO not defined.\n");
-    return( 0 );
+    mbedtls_exit( 0 );
```
Comment (AndrzejKurek) [inline programs/pkey/rsa_verify.c]: I'm not sure if we should return 0 in such cases, but that's probably a remark for a new pr.

File: programs/test/udp_proxy.c [hunk 2]
```diff
@@ -938,7 +939,7 @@ int main( int argc, char *argv[] )
     fflush( stdout ); getchar();
 #endif
 
-    return( exit_code );
+    mbedtls_exit( exit_code );
```
Comment (AndrzejKurek) [inline programs/test/udp_proxy.c]: exit( 1 ) left at line 178/179

**Comment by k-stachowiak** [comment]
@RonEld @AndrzejKurek I have rebased and fixed this PR and its backports. Could you please re-review them?

**Comment by mpg** [comment]
Tests are passing except the API-ABI check which is a false positive: it's noticing a change that happened in development in the meantime, and Mbed OS tests which are an unrelated issue.

The only conflict is in the ChangeLog and can be resolved manually when merging.

**Comment by gilles-peskine-arm** [comment]
Due to the age of the PR, the changelog entry would have been merged into an old release. So I took the liberty of rewriting the history to remove the commit that adds an entry to `ChangeLog`, and instead add an entry with the same text in `ChangeLog.d`. It doesn't matter that this branch doesn't have `ChangeLog.d` yet, the new entry will be picked up after the merge anyway. I did the same thing in the backports.

**Comment by mpg** [comment]
Thanks @gilles-peskine-arm for catching the issue with the ChangeLog. I'll pay more attention to that in the future when reviewing old PRs.

I checked the failures in the pr-head job: check-files and Mbed OS testing. It looks like to me like issues that have been fixed in development in the meantime, which is consistent with the fact that the pr-merge job passed. So IMO it's good enough for merging.

**Pull request merged**